doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,433 @@
+`frames` directly without a copy.
+
+```c
+enum packet_validity_e {
+    QUIC_PACKET_INITIAL,
+    QUIC_PACKET_HANDSHAKE,
+    QUIC_PACKET_0_RTT,
+    QUIC_PACKET_1_RTT
+};
+
+typedef enum packet_validity_e PACKET_VALIDITY;
+
+struct ossl_quic_packet_st {
+    QUIC_CONNECTION *conn;
+    unsigned char *packet_header;
+    size_t packet_header_length;
+    STACK_OF(OSSL_QUIC_TXP_FRAME) *frames;
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+    size_t packet_length;
+
+    /* Pkts - might be better as one field as per ACK Manager */
+    unsigned int initial : 1;
+    unsigned int handshake : 1;
+    unsigned int rtt0 : 1;
+    unsigned int rtt1 : 1;
+    /* Pkts options */
+    PACKET_VALIDITY validity;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` structure from the ACK Manager overlaps this
+  significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in
+  the Application Data number space?
+- 0-RTT packets can morph into 1-RTT packets and this needs to be handled by
+  the packetiser.
+- the `packet_header` will become a structure / union most likely.
+
+Interactions
+------------
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+__owur int ossl_quic_packetise(SSL_QUIC_TX_PACKETISER *tx, QUIC_STREAM *stream,
+                               const OSSL_QUIC_FRAME *frame,
+                               size_t frame_length);
+```
+
+### Stream Send Buffers
+
+Data from the stream send buffers is treated specially.  The packetiser knows
+how much space is left in each packet and it will request that amount of data
+from the stream send buffers.  The stream send buffers will return a
+constructed frame header and a pointer to the steam data and length.
+
+```c
+__owur int ossl_quic_get_app_data(SSL_QUIC_TX_PACKETISER *tx, size_t request,
+                                  const OSSL_QUIC_FRAME **frame,
+                                  const void *data, size_t data_len);
+```
+
+#### Notes
+
+* Unclear how to best free the data after sent data was acked.
+  The data will be fragments from the buffers so the stream send buffers will
+  need to remember which fragment have been sent and which are pending and
+  only free once everything is sent:
+
+```c
+int ossl_quic_free_app_data(OSSL_QUIC_FRAME *frame,
+                            void *data, size_t data_len);
+```
+
```

Comment:
The running total would be the stats manager or ACK managers concern.  I really don't think it belongs here.

It's the ACK manager's job to tell the TXP that retransmission is required & what needs to be retransmitted.  The TXP then rebuilds packets based on this information.  This complicates the coordination between the TXP and the send buffers because data block boundaries can be expected to change.  I've not got a way to say forget about these blocks they need to be retransmitted and the boundaries could change.